packages/babel-core/src/transform-ast.js

```diff
@@ -1,21 +1,56 @@
 // @flow
-import * as t from "@babel/types";
-import loadConfig from "./config";
-import runTransform, { type FileResult } from "./transformation";
-
-export default function transformFromAst(
-  ast: Object,
-  code: string,
-  opts: Object,
-): FileResult | null {
-  const config = loadConfig(opts);
-  if (config === null) return null;
-
-  if (ast && ast.type === "Program") {
-    ast = t.file(ast, [], []);
-  } else if (!ast || ast.type !== "File") {
-    throw new Error("Not a valid ast?");
+
+import loadConfig, { type InputOptions } from "./config";
+import {
+  runAsync,
+  type FileResult,
+  type FileResultCallback,
+} from "./transformation";
+
+import transformAstSync from "./transform-ast-sync";
+
+type AstRoot = BabelNodeFile | BabelNodeProgram;
+
+type TransformAst = {
+  (ast: AstRoot, code: string, callback: FileResultCallback): void,
+  (
+    ast: AstRoot,
+    code: string,
+    opts: ?InputOptions,
+    callback: FileResultCallback,
+  ): void,
+
+  // Here for backward-compatibility. Ideally use ".transformSync" if you want
+  // a synchronous API.
+  (ast: AstRoot, code: string, opts: ?InputOptions): FileResult | null,
+};
+
+export default ((function transformFromAst(ast, code, opts, callback) {
+  if (typeof opts === "function") {
+    opts = undefined;
+    callback = opts;
   }
 
-  return runTransform(config, code, ast);
-}
+  // For backward-compat with Babel 6, we allow sync transformation when
+  // no callback is given. Will be dropped in some future Babel major version.
+  if (callback === undefined) return transformAstSync(ast, code, opts);
+
+  // Reassign to keep Flowtype happy.
+  const cb = callback;
+
+  // Just delaying the transform one tick for now to simulate async behavior
+  // but more async logic may land here eventually.
+  process.nextTick(() => {
```

Comment:
Can `process.nextTick(() => {` be moved inside `runAsync`? So that `runAsync` is actually async and we avoid faking the async behavior two times.

Reply:
I can do that too. I left it this way because I wanted the `cb(err)` for the `loadConfig` call to also be async, and it felt unnecessary to delay things two ticks when one was enough for this case.

Reply:
Oh right... It's ok as is then.